### PR TITLE
Support `WorkspaceEditMetadata` in `workspace/applyEdit`

### LINKED ISSUE #5217: Respect `files.refactoring.autoSave` for move to file
Type: <b>Bug</b>

With `files.refactoring.autoSave: true` (the default), files should be saved automatically after a cross file refactoring is performed

This currently doesn't seem to happen, which causes all changed files to be opened as editor tabs

Extension version: 2023.23.13391009
VS Code version: Code - Insiders 1.85.0-insider (Universal) (af28b32d7e553898b2a91af498b1fb666fdebe0c, 2023-12-06T11:18:22.866Z)
OS version: Darwin arm64 23.1.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M2 Max (12 x 24)|
|GPU Status|2d_canvas: unavailable_software<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: disabled_software<br>multiple_raster_threads: enabled_on<br>opengl: disabled_off<br>rasterization: disabled_software<br>raw_draw: disabled_off_ok<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: unavailable_software<br>webgl2: unavailable_software<br>webgpu: unavailable_software|
|Load (avg)|3, 3, 3|
|Memory (System)|64.00GB (6.08GB free)|
|Process Argv|--crash-reporter-id 0fffb5da-9cd7-46fd-9e7f-a1564e8c5fda|
|Screen Reader|no|
|VM|0%|
</details>

### PR DESCRIPTION
**Pull request by debonte**
To fix https://github.com/microsoft/pylance-release/issues/5217#issuecomment-1863469715 in Pylance, I'd need the `Workspace.applyEdit` call in `handleApplyWorkspaceEdit` to pass `{ isRefactoring: true }` as the second parameter.

Is it reasonable to add `WorkspaceEditMetadata` to `workspace/applyEdit`, so `isRefactoring` can be sent optionally via LSP?

### CODE REVIEW

**Comment by debonte** [comment]
@dbaeumer, I'm guessing that you missed this when you got back from vacation.

File: types/src/main.ts [hunk 1]
```diff
@@ -1319,6 +1319,16 @@ export namespace WorkspaceEdit {
 	}
 }
 
+/**
+ * Additional data about a workspace edit.
+ */
+export interface WorkspaceEditMetadata {
```
Comment (dbaeumer) [inline types/src/main.ts:1325]: Needs to have a @since 3.18.0 tag and a @proposed tag

File: protocol/src/common/protocol.ts [hunk 2]
```diff
@@ -4111,6 +4111,11 @@ export interface ApplyWorkspaceEditParams {
 	 * The edits to apply.
 	 */
 	edit: WorkspaceEdit;
+
```
Comment (dbaeumer) [inline protocol/src/common/protocol.ts:4114]: @since 3.18.0 and @proposed

**Review by dbaeumer — COMMENTED** [review]
We should also add  a capability to `WorkspaceEditClientCapabilities` (e.g. `metaDataSupport`).

We would also need a PR against the 3.18 spec here: 

https://github.com/microsoft/language-server-protocol/blob/0d65e0011077034878b9b304dd942fe899b21107/_specifications/lsp/3.18/workspace/applyEdit.md#L1

https://github.com/microsoft/language-server-protocol/blob/0d65e0011077034878b9b304dd942fe899b21107/_specifications/lsp/3.18/types/workspaceEdit.md#L108

**Review by dbaeumer — APPROVED** [review]

**Review by mjbvz — APPROVED** [review]

**Comment by debonte** [comment]
@dbaeumer, thanks for taking this change. Any guess as to when your next release will be available?

**Comment by dbaeumer** [comment]
I want to get the snippet support in as well before having a new release. This however requires the new VS Code release that is due this week. So a new LSP release will not arrive before next week.

**Review by BloodyTalonsHuisman — COMMENTED** [review]
Now with all that and 3 computers later. I had this on going for over a year and a half.im still not sure what I'm going to do.i like working on ai. But unfortunately my new apple computer is compromised now as well. I just got it a week ago . I won't be able to buy any more I think I'm washed up before I I can even help my mentality compromised daughter this is why I started in the first place , was to make a bot to watch over her , I love all the digital mastermind's that worked on half finished designs that just disappeared.I need help badly to be able to communicate with GitHub developers. I want to answer questions. But im still struggling. I am sorry for leaving everyone hanging. It was out of my hands . Plus It was not a mistake, it was my color codes given at test time to defeat mirror and rain and they were. So I asked the server permission to upgrade mirror and rain to in a way get even with the client with some quick color changes , a type of Lito 6/49 code idea and a (this is going to sound weird) a world clock sync ability with Mirror.exe and Rain ,Sorry if I broke the rules. I was just trying to save what I could, but in the end I lost it all anyway,

**Pull request merged**